python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,267 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Utility module for various conversions around PrimitiveType implementations
+
+This module enables:
+    - Converting partition strings to built-in python objects
+    - Converting a value to a byte buffer
+    - Converting a byte buffer to a value
+
+Note:
+    Conversion logic varies based on the PrimitiveType implementation. Therefore conversion functions
+    are defined here as generic functions using the @singledispatch decorator. For each PrimitiveType
+    implementation, a concrete function is registered for each generic conversion function. For PrimitiveType
+    implementations that share the same conversion logic, registrations can be stacked.
+
+Example (registering `from_partition_to_py` logic for BooleanType):
+    >>> @from_partition_value_to_py.register(BooleanType)
+    >>> def _(primitive_type, partition_value_as_str: str):
+            if partition_value_as_str is None:
+                return False
+            return partition_value_as_str.lower() == "true"
+
+Example (stacking registrations):
+   >>> @from_partition_value_to_py.register(FloatType)
+   >>> @from_partition_value_to_py.register(DoubleType)
+   >>> def _(primitive_type, partition_value_as_str: str):
+           return float(partition_value_as_str)
+"""
+import struct
+import uuid
+from decimal import ROUND_HALF_DOWN, Decimal, localcontext
+from functools import singledispatch
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+
+@singledispatch
+def from_partition_value_to_py(primitive_type, partition_value_as_str: str):
+    """A generic function function which converts a partition string to a python built-in
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        partition_value_as_str(str): A string representation of a partition value
+    """
+    if partition_value_as_str is None:
+        return None
+    raise TypeError(
+        f"Cannot convert partition string to python built-in, type {primitive_type} not supported: '{partition_value_as_str}'"
+    )
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return False
+    return partition_value_as_str.lower() == "true"
+
+
+@from_partition_value_to_py.register(IntegerType)
+@from_partition_value_to_py.register(LongType)
+def _(primitive_type, partition_value_as_str: str):
+    return int(float(partition_value_as_str))
```

Comment:
Do these also need checks for `None`?

Reply:
Converting to float and then to int allows values through that would fail in Java.

I think we want to be strict here as well. If a value has a decimal point and isn't an integer, I think it is better to fail. This is only used in narrow cases anyway. In fact, I thought about saying we should remove these methods entirely. Let's leave them in, but be strict about the conversions.

Reply:
Thanks for catching this! I added some logic to raise if there are any non-zero fractional digits. ([here](https://github.com/apache/iceberg/pull/4263/files#diff-23863ae95c4894ac0025efabd97dd925503dc5a908e386d7c345952e36b9743cR134)):
```py
@from_partition_value_to_py.register(IntegerType)
@from_partition_value_to_py.register(LongType)
def _(primitive_type, partition_value_as_str: str):
    """
    Note: Before attempting to cast the value to string, a validation happens to ensure that there are no fractional digits. For
    example, an invalid value such as "123.45" will be converted to a decimal with digits (1, 2, 3, 4, 5) with an exponent of -2 and the last
    2 digits will be inspected for non-zero values. An example of a valid value is "123.00".
    """
    if partition_value_as_str is None:
        return None
    _, digits, exponent = Decimal(partition_value_as_str).as_tuple()
    if exponent != 0 and any(digits[exponent:]):  # If there are digits to the right of the exponent, raise if any are not 0
        raise ValueError(f"Cannot convert partition value, value cannot have fractional digits for {primitive_type} partition")
    return int(float(partition_value_as_str))
```

Reply:
I think just directly casting to int is a simpler way to do this. When I remove the inner `float` cast and test it, I get `ValueError: invalid literal for int() with base 10: '12.34'`. That seems correct to me. This method is also _not_ recommended for production use, it is more of an internal convenience.